teps/0023-implicit-mapping.md

```diff
@@ -0,0 +1,358 @@
+support remote TaskRefs or PipelineRefs.
+
+### Caveats
+
+#### Naming conflicts
+
+In cases of parameter naming conflicts, the innermost definition should win.
+e.g. for the following config:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: pipelinerun-with-taskspec-to-echo-message
+spec:
+  pipelineSpec:
+    tasks:
+      - name: echo-message
+        taskSpec:
+          params:
+            - name: MESSAGE
+              type: string
+          steps:
+            - name: echo
+              image: ubuntu
+              script: |
+                #!/usr/bin/env bash
+                echo "$(params.MESSAGE)"
+  params:
+    - name: MESSAGE
+      value: ["Good Morning!"]
+```
+
+The innermost Task definition should take precedence and make the config
+**invalid**, since the PipelineRun array param should not be able to override
+the param definition of the task.
+
+#### Explicit Definitions
+
+We should only define implicit params if no params section is defined. This
+ensures that this behavior will remain backwards compatible for existing
+defintions. e.g.
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: pipelinerun-with-taskspec-to-echo-message
+spec:
+  pipelineSpec:
+    params:
+      - name: MESSAGE
+        type: string
+    tasks:
+      - name: echo-message
+        taskSpec:
+          steps:
+            - name: echo
+              image: ubuntu
+              script: |
+                #!/usr/bin/env bash
+                echo "$(params.MESSAGE)"
+  params:
+    - name: MESSAGE
+      value: "Good Morning!"
+    - name: UNUSED
+      value: "unused message"
+```
+
+The `UNUSED` param will _not_ be plumbed through to the underlying embedded Task
+(even though it's an extra parameter to the Pipeline).
+
+#### Common parameter names
+
+Certain parameter names may be relatively common and therefore carry a different
+meaning between different `Tasks` (e.g. `url`, `path`, etc). `Tasks` with the
+same parameter names that require different input could therefore inadvertently
+acquire the wrong value.
+
+We consider it the `Pipeline` author's responsibility to guard against this.
+
+Our expectation here is that if authors are choosing to embed specs, then the
+context of the variables should be known to avoid problems like this. If they
+require a separate/different mapping then it would be up to the author to
+explicitly assign these params, either in line or in its own Task definition.
+
+### User Experience
+
+Our goal with this change is to make it easier to write PipelineRuns. We
+particularly see value in getting started use cases where users may want to
+start with a single-file PipelineRun with embedded specs before refactoring into
+different Pipelines/Tasks for reusability later (similar to how one might write
+a binary with a single `main` before refactoring to different funcs/packages).
```

Comment:
i think it could really help to include a larger, more verbose, and more real world example - e.g. something with a LOT of duplication and verbosity - maybe dogfooding has some examples?

Reply:
Dogfooding is tricky because we rely a lot on TriggerTemplates, and try to reuse components as much as possible (e.g. with the various trigger handlers). In some ways, the fact

There's definitely [examples](https://github.com/tektoncd/plumbing/blob/bd1811ea7d20fe8a9ff41bdbf02ea8583c586e97/tekton/ci-workspace/plumbing/template.yaml) of repetitive configs that I'd like to take a look at after to see how we could improve, but it's out of scope for this proposal since we don't want to dive into TriggerTemplates / Remote resources just yet.